Fix using mouse alleles and added predictor classes

skrakau opened the pull request:
For the nf-core/epitopeprediction pipeline I need the support for mouse alleles and tried to do the required changes myself.
I updated the by MHCNuggets supported MHC class I alleles: just copied all "HLA" and "H2" alleles listed for MHCNuggets v2.3.2. This also removed for example "HLA-A*02:04", since it is not listed by MHCNuggets. As the epitopeprediction pipeline I used version 2.3.2, is that OK? 
 
Additionally, it seems that MHCNuggets expects the mouse alleles to have the format "H-2-Xx", while when reading the allele input file, everything is capitalised. Since the "H-2-Xx" format is also described within the MouseAllele class, I simply changed the internal representation.

Let me know if I should change something.

skrakau commented:
The syfpeithi website states it also supports a few mouse alleles, but I could not see how to do that within FRED2.

skrakau commented:
I will change this to introduce a new class for MHCnuggets version v.2.3.2 with updated supported alleles.

b-schubert commented:
Yes, thanks that would be great. Other than that PR looks good. 

Could you check whether the Alleles in the Results data frame are actually the allele object and not just string representations? As of right now, it seems that they are just strings? The same goes for MHCFlurry....

christopher-mohr commented:
> Additionally, it seems that MHCNuggets expects the mouse alleles to have the format "H-2-Xx", while when reading the allele input file, everything is capitalised. 

Where does that happen? 

I just tried it out and it seems to work fine:

```
>>> allele = Allele("H2-Dd")
>>> allele
H-2-Dd
```

Although I think the available mouse alleles for netMHC and netMHCpan in the corresponding classes seem to be wrong. It should be 'H-2-Db' 'instead of 'H2-Db' e.g.

skrakau commented:
> > Additionally, it seems that MHCNuggets expects the mouse alleles to have the format "H-2-Xx", while when reading the allele input file, everything is capitalised.
> 
> Where does that happen?

when reading alleles with `read_lines()` in https://github.com/FRED-2/Fred2/blob/master/Fred2/IO/FileReader.py. One could also adjust this directly in this function when `in_type` is `Allele`, but I thought it might be better to change it directly in the  `MouseAllele` class, to allow different formats provided by the user.

skrakau commented:
What about alleles such as "HLA-BF02:0401" ? It is listed for MHCnuggets v2.0, but not within FRED2 for MHCnuggets v2.0. Is there a reason for this? 

I will add the mouse alleles also for MHCnuggets v2.0 and am just wondering if I should add anything else or not include such alleles in general.

b-schubert commented:
Generally, ```Allele``` implements the official HLA nomenclature defined by the WHO nomenclature committee:

http://hla.alleles.org/nomenclature/index.html

Every tool that deviates from that nomenclature needs to implement its own conversion function as indicated by the defined interfaces.

skrakau commented:
Thanks @b-schubert!

I was mainly wondering about the "BF" part, but then I would just ignore for now everything different than HLA-A/B/C*NN:NN (for class I).

Is it OK to adjust the lower case for "H-2-Xx" within the `MouseAllele` class directly (and not just turning off the capitalisation within `read_lines()`) to allow for different user inputs?

skrakau commented:
In the meantime I added the following changes:

- added new classes for MHCnuggets v2.3.2 and MHCFlurry v1.4.3
- added mouse alleles ("H-2-") and corresponding conversions to the old MHCnuggets and MHCFlurry versions
- corrected the listed `supported_lengths` for MHCnuggets v2.0 class-1 (causing "IndexError: list index out of range" errors), although it is not yet used to handle those cases
- changed the `revert_allele_repr()` function to return `Allele` instead of `str`
- changed the handling of the MHCnuggets results, because they are not always written out directly due to a not closed filehandle (https://github.com/KarchinLab/mhcnuggets/pull/8), which sometimes causes errors or missing results: 
   - running MHCnuggets without output file and instead capture results written to stdout

However, I don't know why the tests are failing.

christopher-mohr commented:
It seems like something has changed for the dependencies (maybe tensorflow). We will have to find out what changed exactly.

christopher-mohr commented:
I guess the found the cause of this error. They released a new MHCflurry version on pypi on 20th of March: https://pypi.org/project/mhcflurry/#history 

I guess one way to fix this would be to specify the version of MHCflurry.

christopher-mohr commented:
@b-schubert what do you think how we should deal with different MHCflurry/nuggets versions? It's a different situation than with netMHC(pan) since it is installed with FRED2. Which version of MHCflurry should be installed? I guess in case a user wants to use an older version he has to deal with that himself and install the corresponding version in his environment?!

skrakau commented:
Are there any news on this topic? Can the MHCflurry version be set to <1.6 for the beginning to avoid this error?

christopher-mohr commented:
> Are there any news on this topic? Can the MHCflurry version be set to <1.6 for the beginning to avoid this error?

Yes, let's do it and see if the checks pass...

christopher-mohr commented on Fred2/EpitopePrediction/ANN.py:
```diff
@@ -123,15 +140,40 @@ def version(self):
+            :return: str
+            """
+            if isinstance(allele, MouseAllele):
+                return "H-2-%s%s%s" % (allele.locus, allele.supertype.upper(), allele.subtype)
```
I guess you tested if that works? Wasn't sure because of the supertype/subtype thing and the difference to human alleles when I just read that.

skrakau replied:
tested ;)

christopher-mohr commented on Fred2/EpitopePrediction/ANN.py:
```diff
@@ -123,15 +140,40 @@ def version(self):
+            :type alleles: :class:`~Fred2.Core.Allele.Allele`
+            :return: str
+            """
+            if isinstance(allele, MouseAllele):
```
I think we could could even replace "HLA" and "H-2" with `allele.organism`.

skrakau replied:
Changed this in ANN.py, but didn't touch other files.
